storage/consistent_watchable_store.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"log"
+	"strconv"
+)
+
+var (
+	indexBucketName = []byte("index")
+	indexKeyName    = []byte("index")
+)
+
+// IndexGetter is an interface that wraps the Get method.
+type IndexGetter interface {
+	// Get gets the index.
+	Get() uint64
+}
+
+type consistentWatchableStore struct {
+	*watchableStore
+	// The IndexGetter is used to get the index which is consistent with
+	// the on-going txn.
+	// The gotten index is put into the underlying backend along with
+	// the on-going txn. This guarantees that the index is always consistent
+	// with the latest txn, even if program is hard killed at any time.
+	ig IndexGetter
+}
+
+// newConsistentWatchableStore creates a new consistentWatchableStore using the file at the given path.
+// If the file at the given path does not exist then it will be created automatically.
+// The given IndexGetter should return the index which is consistent with
+// the on-going txn.
+func newConsistentWatchableStore(path string, ig IndexGetter) *consistentWatchableStore {
+	s := &consistentWatchableStore{
+		watchableStore: newWatchableStore(path),
+		ig:             ig,
+	}
+
+	b := s.watchableStore.store.b
+	tx := b.BatchTx()
+	tx.Lock()
+	tx.UnsafeCreateBucket(indexBucketName)
+	tx.Unlock()
+	b.ForceCommit()
+
+	return s
+}
+
+func (s *consistentWatchableStore) Put(key, value []byte) (rev int64) {
+	id := s.TxnBegin()
+	rev, err := s.TxnPut(id, key, value)
```

Comment:
I do not really understand why do we need to do this. I guess all we need is to record the latest index.